### PR TITLE
Speedup python local_unitary_simulator by using Numpy.einsum

### LINKED ISSUE #894: Speedup python local_unitary_simulator
@chriseclectic noted in #893

### What is the expected enhancement?
An important speedup by using Einstein summation for enlarging one and two qubit operations.

An example of runtime for a random circuit of 100 gates on my laptop (Macbook pro 3.1 GHz i7, 16 GB Ram):

Timing results for new implementation are:

```
2 qubit circuit, 100 gates, time taken: 0.073 s
3 qubit circuit, 100 gates, time taken: 0.072 s
4 qubit circuit, 100 gates, time taken: 0.071 s
5 qubit circuit, 100 gates, time taken: 0.075 s
6 qubit circuit, 100 gates, time taken: 0.087 s
7 qubit circuit, 100 gates, time taken: 0.08 s
8 qubit circuit, 100 gates, time taken: 0.11 s
9 qubit circuit, 100 gates, time taken: 0.352 s
10 qubit circuit, 100 gates, time taken: 1.346 s
11 qubit circuit, 100 gates, time taken: 4.931 s
12 qubit circuit, 100 gates, time taken: 28.891 s
```

Timing results for old implementation are

```
2 qubit circuit, 100 gates, time taken: 0.24 s
3 qubit circuit, 100 gates, time taken: 0.123 s
4 qubit circuit, 100 gates, time taken: 0.178 s
5 qubit circuit, 100 gates, time taken: 0.114 s
6 qubit circuit, 100 gates, time taken: 0.138 s
7 qubit circuit, 100 gates, time taken: 0.238 s
8 qubit circuit, 100 gates, time taken: 0.486 s
9 qubit circuit, 100 gates, time taken: 2.835 s
10 qubit circuit, 100 gates, time taken: 15.714 s
11 qubit circuit, 100 gates, time taken: 115.502 s
12 qubit circuit, 100 gates, time taken: ?
```
(I wasn't patient enough for the 12 qubit circuit)

The python file used for printing benchmarks was:

```python
# Simple benchmark
import numpy as np
import time
import qiskit

# Create a test circuit
for num_qubits in range(2, 13):
    # Create random test circuit
    num_gates = 100
    qr = qiskit.QuantumRegister(num_qubits)
    circ = qiskit.QuantumCircuit(qr)
    for j in range(num_gates):
        q = np.random.randint(0, num_qubits)
        r = np.random.randint(0, 4)
        if r == 0:       
            circ.cx(qr[q], qr[(q + 1) % num_qubits])
        elif r == 1:
            circ.h(qr[q])
        elif r == 2:
            circ.s(qr[q])
        elif r == 3:
            circ.t(qr[q])
    # Execute with timer
    start = time.time()
    result = qiskit.execute(circ, 'local_unitary_simulator').result()
    stop = time.time()
    print("{} qubit circuit, {} gates, time taken: {} s".format(num_qubits, num_gates, np.round(stop-start,3)))
```

### PR DESCRIPTION
**Pull request by chriseclectic**
Fix #894

### Summary

This PR changes the local unitary simulator to use the Numpy.einsum function. This removes the need to enlarge single and two qubit gates to full 2^N x 2^N matrices, and greatly speeds up performance for larger circuits.

### Details and comments

An example of runtime for a random circuit of 100 gates on my laptop (Macbook pro 3.1 GHz i7, 16 GB Ram):

Timing results for new implementation are:

```
2 qubit circuit, 100 gates, time taken: 0.073 s
3 qubit circuit, 100 gates, time taken: 0.072 s
4 qubit circuit, 100 gates, time taken: 0.071 s
5 qubit circuit, 100 gates, time taken: 0.075 s
6 qubit circuit, 100 gates, time taken: 0.087 s
7 qubit circuit, 100 gates, time taken: 0.08 s
8 qubit circuit, 100 gates, time taken: 0.11 s
9 qubit circuit, 100 gates, time taken: 0.352 s
10 qubit circuit, 100 gates, time taken: 1.346 s
11 qubit circuit, 100 gates, time taken: 4.931 s
12 qubit circuit, 100 gates, time taken: 28.891 s
```

Timing results for old implementation are

```
2 qubit circuit, 100 gates, time taken: 0.24 s
3 qubit circuit, 100 gates, time taken: 0.123 s
4 qubit circuit, 100 gates, time taken: 0.178 s
5 qubit circuit, 100 gates, time taken: 0.114 s
6 qubit circuit, 100 gates, time taken: 0.138 s
7 qubit circuit, 100 gates, time taken: 0.238 s
8 qubit circuit, 100 gates, time taken: 0.486 s
9 qubit circuit, 100 gates, time taken: 2.835 s
10 qubit circuit, 100 gates, time taken: 15.714 s
11 qubit circuit, 100 gates, time taken: 115.502 s
12 qubit circuit, 100 gates, time taken: ?
```
(I wasn't patient enough for the 12 qubit circuit)

The python file used for printing benchmarks was:

```python
# Simple benchmark
import numpy as np
import time
import qiskit

# Create a test circuit
for num_qubits in range(2, 13):
    # Create random test circuit
    num_gates = 100
    qr = qiskit.QuantumRegister(num_qubits)
    circ = qiskit.QuantumCircuit(qr)
    for j in range(num_gates):
        q = np.random.randint(0, num_qubits)
        r = np.random.randint(0, 4)
        if r == 0:       
            circ.cx(qr[q], qr[(q + 1) % num_qubits])
        elif r == 1:
            circ.h(qr[q])
        elif r == 2:
            circ.s(qr[q])
        elif r == 3:
            circ.t(qr[q])
    # Execute with timer
    start = time.time()
    result = qiskit.execute(circ, 'local_unitary_simulator').result()
    stop = time.time()
    print("{} qubit circuit, {} gates, time taken: {} s".format(num_qubits, num_gates, np.round(stop-start,3)))
```

### CODE REVIEW

**Comment by delapuente** [comment]
Remember to open an issue first and link it with "Fix <issuenumber>". I did it for you this time.

**Comment by ajavadia** [comment]
This looks great. It was really slow to enlarge every small gate before. I also like the revamp of tests. Once the linter passes, I think it can be merged.

**Comment by jaygambetta** [comment]
I agree

File: qiskit/backends/local/unitary_simulator_py.py [hunk 1]
```diff
@@ -128,19 +130,60 @@ def _add_unitary_single(self, gate, qubit):
...
         number_of_qubits is the number of qubits in the system.
         """
-        unitary_add = enlarge_single_opt(gate, qubit, self._number_of_qubits)
-        self._unitary_state = np.dot(unitary_add, self._unitary_state)
 
-    def _add_unitary_two(self, gate, q_0, q_1):
+        # Convert to complex rank-2 tensor
+        gate_tensor = np.array(gate, dtype=complex)
+
+        # Compute einsum index string
```
Comment (delapuente) [inline qiskit/backends/local/unitary_simulator_py.py]: Non-blocking: Einstein's summation convention seems really powerful! Could you extract these index calculations into a couple of functions and provide short explanations. A super-brief introduction of the `einsum` function would be great for the coders, like me, that are not experts in the matter. But I'm not blocking this, I understand this is a very specific code.

**Pull request merged**